From b8b19dc7aa2b903e882f29ca5441895a9125d3d1 Mon Sep 17 00:00:00 2001
From: Hendrik Brueckner <brueckner@redhat.com>
Date: Fri, 12 Dec 2014 13:11:31 -0500
Subject: [s390] crypto: kernel oops at insmod of the z90crypt device driver

Message-id: <1418389892-28466-1-git-send-email-brueckner@redhat.com>
Patchwork-id: 101716
O-Subject: [RHEL7.1 PATCH] [s390] crypto: kernel oops at insmod of the z90crypt device driver
Bugzilla: 1172136
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Description
-----------
commit 099eae11a8beb13e17f6b9371c7c090c61bc061d
Author: Harald Freudenberger <freude@linux.vnet.ibm.com>
Date:   Wed Dec 10 10:23:59 2014 +0100

    s390/zcrypt: kernel oops at insmod of the z90crypt device driver

    Kernel oops caused by invalid parameter at TAPQ instruction:
    On older systems where the QCI instruction is not available
    all possible domains are probed via TAPQ instruction. The
    range for the probe has been extended with the > 16 domain
    support now leading to a possible specification exception
    when this instruction is called for probing higher values
    within the new range. This may happen during insmod and/or
    ap bus reset only on machines without a QCI instruction (z10,
    z196, z114), zEC12 and newer systems are not affected.
    The fix modifies the domain checking function to limit the
    allowed range if no QCI info is available.

    Signed-off-by: Harald Freudenberger <freude@linux.vnet.ibm.com>
    Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>

Bugzilla
--------
BZ 1172136
https://bugzilla.redhat.com/show_bug.cgi?id=1172136

Upstream status of the patch
----------------------------
https://git.kernel.org/cgit/linux/kernel/git/next/linux-next.git/commit/?id=099eae11a8beb13e17f6b9371c7c090c61bc061d

Brew
----
https://brewweb.devel.redhat.com/taskinfo?taskID=8369399

Test status
-----------
The patch has been tested and fixes the problem.
The fix has been verified by the IBM test department.

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/s390/crypto/ap_bus.c b/drivers/s390/crypto/ap_bus.c
index d145608..8c2beb9 100644
--- a/drivers/s390/crypto/ap_bus.c
+++ b/drivers/s390/crypto/ap_bus.c
@@ -1157,9 +1157,13 @@ static inline int ap_test_config_card_id(unsigned int id)
  */
 static inline int ap_test_config_domain(unsigned int domain)
 {
- if (!ap_configuration)
-  return 1;
- return ap_test_config(ap_configuration->aqm, domain);
+ if (!ap_configuration)   /* QCI not supported */
+  if (domain < 16)
+   return 1; /* then domains 0...15 are configured */
+  else
+   return 0;
+ else
+  return ap_test_config(ap_configuration->aqm, domain);
 }
 
 /**
-- 
1.7.1